### PR TITLE
VACMS-15798 / 15791 / 15795 next preview configs + oauth key handling

### LINKED ISSUE #15791: Set up basic Next.js site entity
## Requirements

The Next Drupal setup requires a Next.js 'site' entity, so that Drupal knows where its preview server is and what secret to use when accessing the preview server.

```[tasklist]
### Acceptance criteria
- [ ] A Next.js 'site' config entity is present and viewable on all environments
```

## Background & implementation details

Next.js 'site' entites are created through the admin here: /admin/config/services/next/sites/add

These are config entities, so the output of the ticket should be a config entity. 

Label: 'Next Build preview server'
Base url: can be anything; this will ultimately be set in settings.php for the given environment

everything else blank.

```[tasklist]
### Implementation tasks
- [ ] Use this if it helps you or feel free to delete.
```

### LINKED ISSUE #15792: Next.js 'site' configuration for Tugboat
## Requirements

We need specific pieces of information set up for Next Build preview for Tugboat, so that Tugboat CMS can use its corresponding preview server.

```[tasklist]
### Acceptance criteria
- [ ] A `DRUPAL_PREVIEW_SECRET` is set for Tugboat and stored in AWS SSM
- [ ] An environment variable is set on Tugboat for the location of the Preview server
- [ ] The secret and URLs are assigned to the Tugboat Next.js 'site' config via settings.tugboat.php
```

## Background & implementation details

* `DRUPAL_PREVIEW_SECRET` needs to be set and stored in AWS SSM. This is a text value; it can be anything.
* The preview server URL should be set as an environment variable for the Tugboat instance. See this file in the CMS Tugboat setup: https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/.tugboat/.env.j2#L8
* The preview server URL should be assigned to both the Base URL and Preview URL in settings.tugboat.config
* The secret should be read from SSM and assigned appropriate in [settings.tugboat.config](https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/docroot/sites/default/settings/settings.tugboat.php)
* Revalidation settings can be left empty for the time being.
 

```[tasklist]
### Implementation tasks
- [ ] Use this if it helps you or feel free to delete.
```

### LINKED ISSUE #15795: Write Next Build Preview Oauth keys to Tugboat filesystem
## Requirements

We need Oauth keys available for use on Tugboat filesystems, so that they can be used for CMS / Next Build preview authentication.

```[tasklist]
### Acceptance criteria
- [ ] https://github.com/department-of-veterans-affairs/va.gov-cms/issues/15794
- [ ] The Tugboat Oauth keys are read from AWS SSM and written the Tugboat file server outside the docroot
```

## Background & implementation details

These could be set as global, repo-wide environment variables here: https://tugboat.vfs.va.gov/5fd3b8ee7b4657022b5722d6/settings/

The actual values could be set there, so not actually read directly from SSM but still stored there.

Once the env variables exist, we can write the value to file as part of Tugboat setup. This would mean setting this in the `init` phase, and then removing the base preview so that it can be recreated with the values in place.

These values *must* be written to locations outside of docroot.

```[tasklist]
### Implementation tasks
- [ ] Use this if it helps you or feel free to delete.
```

### LINKED ISSUE #15798: Specific Oauth private/public key location in Oauth settings
## Requirements

CMS Oauth config needs to know where pubilc & private key files are, so that it can use those keys for authentication. 

```[tasklist]
### Acceptance criteria
- [ ] CMS Oauth config is updated to point to the location(s) of our Oauth public & private key files.
```

## Background & implementation details

In #15794 and #15795 and #15796, Oauth public/private key pairs were created, stored in AWS SSM, and then written to the file systems of Tugboat / BRD

```[tasklist]
### Implementation tasks
- [ ] Use this if it helps you or feel free to delete.
```

### PR DESCRIPTION
**Pull request by tjheffner**
## Description
Closes #15798. Closes #15791. Closes #15795. Closes #15792.

This is the CMS-half of the work for the above tickets, the next-build specific work was done in: https://github.com/department-of-veterans-affairs/next-build/pull/227. Certain values have been stored in AWS SSM & are made available through env vars set in the Tugboat UI settings.

With all of that in place, this PR can now be tested on CMS Tugboats.

## Testing done
I tested locally with this companion PR: https://github.com/department-of-veterans-affairs/next-build/pull/227

Visit any story page (ex: https://pr15936-r38pi1elxdywfriwbyojfsw3jhrw8kwk.ci.cms.va.gov/wilmington-health-care/stories/wilmington-va-has-highest-drug-take-back-in-eastern-market) and see that the editor experience is unchanged. You can check other content types too, but it is not necessary.

To confirm all of the underlying wiring is in place correctly: enabled `Story` and `Story Listing` as next site entities here: https://pr15936-r38pi1elxdywfriwbyojfsw3jhrw8kwk.ci.cms.va.gov/admin/config/services/next/entity-types

Screenshot 2023-11-08 at 2 40 19 PM

This configuration was purposely not committed to prevent next-build being used for preview until it is ready. A follow-up PR will include these configs + feature flag handling around preview, along with the correct UI (preview link vs iframe takeover).

After the entities are enabled, visit the story page (ex: https://pr15936-r38pi1elxdywfriwbyojfsw3jhrw8kwk.ci.cms.va.gov/wilmington-health-care/stories/wilmington-va-has-highest-drug-take-back-in-eastern-market) and check the view tab. There should be an iframe instead of the standard editor experience. It may take a moment to render, and will likely appear broken until vets-website assets are in place. This is expected.

Now edit that story with some changes and save it as a draft. There should now be a `latest version` tab. Click that (https://pr15936-r38pi1elxdywfriwbyojfsw3jhrw8kwk.ci.cms.va.gov/node/63185/latest) and see that the iframe exists, rendering the updated draft content.

Now create a new story with any data and save it unpublished. This content should also render in the iframe when looking at the `View` tab. (https://pr15936-r38pi1elxdywfriwbyojfsw3jhrw8kwk.ci.cms.va.gov/amarillo-health-care/stories/a-new-story)

You will also see logs from the running next server in the tugboat dashboard here:
https://tugboat.vfs.va.gov/654196858b633cb05e2b7869

## Screenshots
Unchanged:
Screenshot 2023-11-08 at 8 49 33 PM

**AFTER NEXT JS ENTITY IS ENABLED**

Published:
Screenshot 2023-11-08 at 8 50 56 PM
Draft revision:
Screenshot 2023-11-08 at 8 52 27 PM

Draft never published:
Screenshot 2023-11-08 at 8 55 17 PM

## QA steps

What needs to be checked to prove this works?
What needs to be checked to prove it didn't break any related things?
What variations of circumstances (users, actions, values) need to be checked?

- [ ] The editor experience should be unchanged by default. Check this first.
- After enabling next for specific content types (Story), these cases should be checked:
  - [ ] Published
  - [ ] Unpublished draft of previously published content
  - [ ] New draft content (never published)

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [x] Automated tests have passed.
- [x] Code Quality Tests have passed.
- [x] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.
- [x] If there are field changes, front end output has been thoroughly checked.

### Select Team for PR review

- [x] `CMS Team`
- [ ] `Public websites`
- [ ] `Facilities`
- [ ] `User support`
- [x] `Accelerated Publishing`

### CODE REVIEW

File: config/sync/next.next_site.next_build_preview_server.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+uuid: 75dd1bfb-8507-4981-90b6-3830400e1434
+langcode: en
+status: true
+dependencies: {  }
+id: next_build_preview_server
+label: 'Next Build Preview Server'
+base_url: 'http://localhost:3000'
+preview_url: 'http://localhost:3000/api/preview'
```
Comment (tjheffner) [inline config/sync/next.next_site.next_build_preview_server.yml:8]: noting these two urls work for local testing with next-build running separately outside of ddev.
these are overwritten in `settings.tugboat.php`

File: config/sync/simple_oauth.settings.yml [hunk 2]
```diff
@@ -3,7 +3,9 @@ _core:
 access_token_expiration: 300
 authorization_code_expiration: 300
 refresh_token_expiration: 1209600
-remember_clients: true
 token_cron_batch_size: 0
+public_key: ../public.key
+private_key: ../private.key
```
Comment (tjheffner) [inline config/sync/simple_oauth.settings.yml:8]: relative to `docroot/`

File: .tugboat/config.yml [hunk 3]
```diff
@@ -104,6 +104,10 @@ services:
         - cp "${TUGBOAT_ROOT}"/.tugboat/*.crt /usr/local/share/ca-certificates
         - update-ca-certificates
 
+        # Write OAuth keys to files for next-build Preview Server validation
+        - echo ${OAUTH_PUBLIC_KEY} >> ${TUGBOAT_ROOT}/public.key
```
Comment (ndouglas) [inline .tugboat/config.yml:108]:
```suggestion
        - echo "${OAUTH_PUBLIC_KEY}" >> ${TUGBOAT_ROOT}/public.key
```

File: .tugboat/config.yml [hunk 4]
```diff
@@ -104,6 +104,10 @@ services:
...
+        - echo ${OAUTH_PRIVATE_KEY} >> ${TUGBOAT_ROOT}/private.key
```
Comment (ndouglas) [inline .tugboat/config.yml:109]:
```suggestion
        - echo "${OAUTH_PRIVATE_KEY}" >> ${TUGBOAT_ROOT}/private.key
```

**Review by ndouglas — APPROVED** [review]

**Pull request merged**